jaxlib/cuda_lu_pivot_kernels.cu.cc

```diff
@@ -0,0 +1,95 @@
+/* Copyright 2019 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "jaxlib/cuda_lu_pivot_kernels.h"
+
+#include <array>
+#include <iostream>
+
+#include "jaxlib/cuda_gpu_kernel_helpers.h"
+#include "jaxlib/kernel_helpers.h"
+
+namespace jax {
+namespace {
+
+__device__ void ComputePermutation(const std::int32_t* pivots,
+                                   std::int32_t* permutation_out,
+                                   const std::int32_t num_pivots,
+                                   const std::int32_t num_permutation_rows) {
+  for (int i = 0; i < num_permutation_rows; ++i) {
+    permutation_out[i] = i;
+  }
+
+  // Compute the permutation from a sequence of transpositions encoded in the
+  // pivot array by applying the transpositions in order on the identity
+  // permutation.
+  for (int i = 0; i < num_pivots; ++i) {
+    if (pivots[i] < 0) {
```

Comment:
I'd be tempted to skip out of bounds pivots on the high end also. This should never happen, but it's probably better to avoid corrupting GPU memory in that case.